proposals/p3407.md

```diff
@@ -0,0 +1,191 @@
+        the namespace itself makes lifetimes clearer.
+
+## Alternatives considered
+
+### Allow prefixing a tuple binding pattern with a namespace
+
+We could use the namespace to prefix the binding tuple. For example:
+
+```carbon
+var NS.(a: i32, b: i32) = (3, 4);
+```
+
+It's rare that we would have a single statement declare multiple names. As a
+consequence, the separation of the namespace qualifier from the declared
+identifier might end up unique to this syntax. In that context, we prefer `NS.a`
+for consistency with other cases, such as `class NS.class`.
+
+### Allow binding patterns to declare names in multiple namespaces
+
+We could allow binding patterns to declare names in multiple namespaces. For
+example:
+
+```carbon
+namespace NS;
+var (NS.a: i32, b: i32) = InitData();
+```
+
+Mixing namespaces could be confusing: for example, `b` could be misunderstood to
+be declared in `NS`. We lack data that would demonstrate benefits to offset
+that.
+
+We disallow mixing namespaces in a single declaration for simplicity.
+
+### Allow declaring names in namespaces not owned by the current scope
+
+We could allow declaring names in namespaces not owned by the current scope. For
+example:
+
+```carbon
+namespace NS;
+class ClassT {
+  var NS.val: i32;
+
+  class NS.ChildT {}
+}
+```
+
+Here, `package.NS.val` would be a global, but `ClassT.NS.val` looks more like an
+instance member. It's also unclear whether `ClassT.NS.val` (or
+`instance.NS.val`) could be used to reference the produced variable, since `NS`
+is not inside `ClassT`'s name scope. The naming problems extend to non-binding
+declarations such as `NS.ChildT`, too.
+
+Disallowing using namespaces to cross name scopes is consistent with rules that
+generally prevent declaring names in other name scopes, such as:
+
+```carbon
+class A {
+    class B {
+        // `C` must be declared directly inside `A`.
+        class A.C;
+    }
+}
+
+// `D` must be declared within `A`, even if separately defined.
+class A.D {}
+```
+
+Both the `namespace` declaration and names declared within the `namespace` must
+be written in the same name scope. This avoids name lookup ambiguities, and
+builds consistency in name scope boundaries across declarations.
+
+### Disallow declaring namespaces other than the file scope
+
+We could disallow declaring namespaces in scopes other than the file scope.
+Combined with requiring members of namespaces to only be declared at the file
+scope, this would prevent their use in other contexts.
+
+This would be consistent with C++. However, we believe the current design
+doesn't actively interfere with Carbon, and it could prove useful in some
+situations. For example, perhaps a complex class would find a member namespace
+useful:
+
+```carbon
+class Complex {
+   namespace OptionSet1;
+   class OptionSet1.MemberClassA;
+   class OptionSet1.MemberClassB;
+
+   namespace OptionSet2;
+   class OptionSet2.MemberClassC;
+   class OptionSet2.MemberClassD;
+}
```

Comment:
I think there are language semantics issues here that go beyond the implementation -- if we have `x: Complex`, what is the type of `x.OptionSet1`? If it can have instance members, then is it a class? Or is it some new kind of class-namespace-fragment thing? Or something else? Is it a first-class value that can be passed around, or like a namespace can it only be used in limited ways?

To be clear: I think these questions can be answered. But I think we shouldn't add the functionality without answering them, and I don't think we should block this proposal on figuring out the answers :) I'm very happy for you to be setting this to the side for now.